satpy/readers/yaml_reader.py

```diff
@@ -990,11 +992,16 @@ def _find_missing_segments(file_handlers, ds_info, dsid):
     failure = True
     counter = 1
     expected_segments = 1
+    # get list of file handlers in segment order
+    # (ex. first segment, second segment, etc)
     handlers = sorted(file_handlers, key=lambda x: x.filename_info.get('segment', 1))
     projectable = None
     for fh in handlers:
         if fh.filetype_info['file_type'] in ds_info['file_type']:
             expected_segments = fh.filetype_info.get('expected_segments', 1)
+            if isinstance(expected_segments, str):
+                # get the information from the filename
+                expected_segments = fh.filename_info[expected_segments]
```

Comment:
This is the same code as above, could we factorize it, like eg by replacing the value of `expected_segments` in the filetype info ?

Reply:
Yes, I thought about that and tried it but my initial attempt at doing that didn't work. I could at the very least put this in a separate function of sorts.

As far as "parsing the yaml a second time", I'm not sure I understand what you mean. If the expected_segments is a string then we get the information from the filename rather than an integer. The YAML isn't parsed again.

Reply:
True, I was just exaggerating a bit :)

Reply:
The other possibility, but less flexible, is to assume that if `expected_segments` isn't defined that we should look for `total_segments` in the filename info. If that isn't defined then default to 1 as it does now.